### ISSUE TITLE
Make Appium great with OpenCV

### ISSUE THREAD

**Issue by mykola-mokhnach** [appium/appium#10545]
Integrate opencv4nodejs library into appium-support module, so we could use OpenCV-backed picture matching and comparison features.

**Comment by mykola-mokhnach** [comment]
Linking https://github.com/appium/appium-support/pull/65

**Comment by vjmaiky** [comment]
when it will be available for usage?

**Comment by mykola-mokhnach** [comment]
As soon as @jlipps comes up with API draft, where these helpers can be used

**Comment by mykola-mokhnach** [comment]
Linking https://github.com/appium/java-client/pull/870

**Comment by vjmaiky** [comment]
@mykola-mokhnach even if this thread is closed i have a question regarding it.
I looked over the documentation that you provided:

> +- OpenCV 3+ native libraries
> +- [opencv4nodejs](https://github.com/justadudewhohacks/opencv4nodejs) npm module: `npm i -g opencv4nodejs`. By default the preinstall script of this module also downloads and makes all the required OpenCV libs from source, but this requires developer tools to be available on the host system.
> +- Appium Server 1.8.0+

Still i cannot make it work. How did you added that repository as dependency on your java project? (i downloaded a jar 3.2.0 but it doesn't contain the methods from your example).

Can you give me a hint here?

**Comment by mykola-mokhnach** [comment]
jars are for Java and you need native libs (aka .dll or .so)

**Comment by vjmaiky** [comment]
let me give and example, because it is not clear for me:

My automation project is java on gradle structure

How i can add openCV libraries (.dll or .so) as dependencies to my project? or how i can use them?

If i use your snnipet code, i cannot import OcurrenceMatchingResult or FeaturesMatchingResult, etc:

```
byte[] screenshot = Base64.encodeBase64(driver.getScreenshotAs(OutputType.BYTES));
OccurrenceMatchingResult result = driver
        .findImageOccurrence(screenshot, partialImage, new OccurrenceMatchingOptions()
                .withEnabledVisualization());
assertThat(result.getVisualization().length, is(greaterThan(0)));
assertNotNull(result.getRect());
```

**Comment by mykola-mokhnach** [comment]
java primitives, like `OccurrenceMatchingResult` one  are only accessible in the recent BETA version of the client. 

Regarding OpenCV - simply run `npm i -g opencv4nodejs` on the server machine where Appium is running and everything should be done after the installation succeeds.

**Comment by vjmaiky** [comment]
other version than [Appium] Welcome to Appium v1.8.1-beta1 ?

**Comment by suswari** [comment]
Hi @mykola-mokhnach ,
I am trying to use appium image comparison and used the reference code 
byte[] screenshot = Base64.encodeBase64(driver.getScreenshotAs(OutputType.BYTES));
OccurrenceMatchingResult result = driver
        .findImageOccurrence(screenshot, partialImage, new OccurrenceMatchingOptions()
                .withEnabledVisualization());
assertThat(result.getVisualization().length, is(greaterThan(0)));
assertNotNull(result.getRect());

I see the following error:
org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: opencv4nodejs module is required to use OpenCV features. Please install it first (npm i -g opencv4nodejs) and restart Appium. Read https://github.com/justadudewhohacks/opencv4nodejs#how-to-install for more details on this topic.

My project is in java and I am really not sure about performing npm i -g opencv4nodejs. Should I do npm i -g opencv4nodejs under the project root, though the project doest not have package.json

**Comment by vkuhi** [comment]
@suswari How did you solve this issue?

**Comment by KazuCocoa** [comment]
Did you install opencv4nodejs reading the link if you cannot install it via `npm i -g opencv4nodejs` ?
Appium requires the library to do the feature as a node module.

**Comment by rockzill** [comment]
> Did you install opencv4nodejs reading the link if you cannot install it via `npm i -g opencv4nodejs` ?
> Appium requires the library to do the feature as a node module.

Yes, but jar is not generated and I don't know why :( is it possible to reuse someone's else JAR lib in my project? May be you can share yours and I can add it in my project?
